File: firebase/src/normalization.ts
import crypto from "node:crypto";

const COMPANY_SUFFIXES = /\b(inc|incorporated|llc|ltd|limited|corp|corporation|co|company)\b/g;

export function normalizeCompanyName(name: string): string {
  return name
    .toLowerCase()
    .replace(/&/g, " and ")
    .replace(COMPANY_SUFFIXES, "")
    .replace(/[^a-z0-9]+/g, " ")
    .trim()
    .replace(/\s+/g, " ");
}

export function compactName(name: string): string {
  return normalizeCompanyName(name).replace(/\s+/g, "");
}

export function canonicalDomain(rawUrl: string): string | null {
  try {
    const parsed = new URL(rawUrl);
    const host = parsed.hostname.toLowerCase().replace(/^www\./, "");
    return isSharedAtsHost(host) ? null : host;
  } catch {
    return null;
  }
}

export function atsSlug(rawUrl: string): string | null {
  try {
    const parsed = new URL(rawUrl);
    const host = parsed.hostname.toLowerCase().replace(/^www\./, "");
    const segments = parsed.pathname.split("/").filter(Boolean);
    if ((host === "boards.greenhouse.io" || host === "job-boards.greenhouse.io") && segments[0]) {
      return `greenhouse:${segments[0].toLowerCase()}`;
    }
    if (host === "jobs.lever.co" && segments[0]) {
      return `lever:${segments[0].toLowerCase()}`;
    }
    if (host === "jobs.ashbyhq.com" && segments[0]) {
      return `ashby:${segments[0].toLowerCase()}`;
    }
    if (host.endsWith(".ashbyhq.com")) {
      return `ashby:${host.split(".")[0]}`;
    }
    if (host.endsWith(".workdayjobs.com")) {
      return `workday:${host.split(".")[0]}`;
    }
    return null;
  } catch {
    return null;
  }
}

export function canonicalRoleUrl(rawUrl: string): string {
  const parsed = new URL(rawUrl);
  const ignoredParams = new Set([
    "ref",
    "source",
    "utm_campaign",
    "utm_content",
    "utm_medium",
    "utm_source",
    "utm_term",
  ]);
  for (const param of [...parsed.searchParams.keys()]) {
    if (ignoredParams.has(param.toLowerCase())) {
      parsed.searchParams.delete(param);
    }
  }
  parsed.hash = "";
  parsed.pathname = parsed.pathname.replace(/\/+$/, "") || "/";
  parsed.searchParams.sort();
  return parsed.toString();
}

export function roleIdentity(rawUrl: string, postingId?: string | null): string | null {
  const parsed = new URL(rawUrl);
  const host = parsed.hostname.toLowerCase().replace(/^www\./, "");
  const segments = parsed.pathname.split("/").filter(Boolean);
  const board = atsSlug(rawUrl);

  if (host === "boards.greenhouse.io" || host === "job-boards.greenhouse.io") {
    const jobsIndex = segments.findIndex((segment) => segment.toLowerCase() === "jobs");
    const urlPostingId = jobsIndex >= 0 ? segments[jobsIndex + 1] : null;
    const greenhouseId = urlPostingId || parsed.searchParams.get("gh_jid") || postingId;
    return board && greenhouseId ? `${board}:job:${greenhouseId}` : null;
  }

  if (host === "jobs.ashbyhq.com" || host.endsWith(".ashbyhq.com")) {
    const postingSegment = host === "jobs.ashbyhq.com" ? segments[1] : segments[0];
    const ashbyId = postingSegment || postingId;
    return board && ashbyId ? `${board}:job:${ashbyId.toLowerCase()}` : null;
  }

  if (host === "jobs.lever.co") {
    const leverId = segments[1] || postingId;
    return board && leverId ? `${board}:job:${leverId.toLowerCase()}` : null;
  }

  if (postingId) {
    return `posting:${host}:${postingId.toLowerCase()}`;
  }

  const canonical = canonicalRoleUrl(rawUrl);
  return isGenericCareerPath(parsed.pathname) ? null : `url:${canonical}`;
}

export function isSharedAtsHost(host: string): boolean {
  return host === "boards.greenhouse.io" ||
    host === "job-boards.greenhouse.io" ||
    host === "jobs.ashbyhq.com" ||
    host === "jobs.lever.co";
}

function isGenericCareerPath(pathname: string): boolean {
  const normalized = pathname.replace(/\/+$/, "").toLowerCase();
  return normalized === "" || normalized === "/" ||
    /\/(careers?|jobs?|opportunities)$/.test(normalized);
}

export function stableId(prefix: string, value: string): string {
  const hash = crypto.createHash("sha256").update(value).digest("hex").slice(0, 24);
  return `${prefix}_${hash}`;
}

File: src/callumployed/central/normalization.py
from __future__ import annotations

from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

_TRACKING_PARAMETERS = {
    "ref",
    "source",
    "utm_campaign",
    "utm_content",
    "utm_medium",
    "utm_source",
    "utm_term",
}
_SHARED_ATS_HOSTS = {
    "boards.greenhouse.io",
    "job-boards.greenhouse.io",
    "jobs.ashbyhq.com",
    "jobs.lever.co",
}


def canonical_company_domain(raw_url: str) -> str | None:
    host = _hostname(raw_url)
    return None if not host or host in _SHARED_ATS_HOSTS else host


def ats_slug(raw_url: str) -> str | None:
    parsed = urlsplit(raw_url)
    host = (parsed.hostname or "").lower().removeprefix("www.")
    segments = [segment.lower() for segment in parsed.path.split("/") if segment]
    if host in {"boards.greenhouse.io", "job-boards.greenhouse.io"} and segments:
        return f"greenhouse:{segments[0]}"
    if host == "jobs.ashbyhq.com" and segments:
        return f"ashby:{segments[0]}"
    if host.endswith(".ashbyhq.com") and host != "jobs.ashbyhq.com":
        return f"ashby:{host.split('.')[0]}"
    if host == "jobs.lever.co" and segments:
        return f"lever:{segments[0]}"
    if host.endswith(".workdayjobs.com"):
        return f"workday:{host.split('.')[0]}"
    return None


def canonical_role_url(raw_url: str) -> str:
    parsed = urlsplit(raw_url)
    retained = [
        (key, value)
        for key, value in parse_qsl(parsed.query, keep_blank_values=True)
        if key.lower() not in _TRACKING_PARAMETERS
    ]
    retained.sort(key=lambda item: (item[0].lower(), item[0], item[1]))
    path = parsed.path.rstrip("/") or "/"
    return urlunsplit((parsed.scheme.lower(), parsed.netloc.lower(), path, urlencode(retained), ""))


def role_identity(raw_url: str, posting_id: str | None = None) -> str | None:
    parsed = urlsplit(raw_url)
    host = (parsed.hostname or "").lower().removeprefix("www.")
    segments = [segment for segment in parsed.path.split("/") if segment]
    board = ats_slug(raw_url)

    if host in {"boards.greenhouse.io", "job-boards.greenhouse.io"}:
        lowered = [segment.lower() for segment in segments]
        jobs_index = lowered.index("jobs") if "jobs" in lowered else -1
        url_posting_id = segments[jobs_index + 1] if jobs_index + 1 < len(segments) else None
        query = dict(parse_qsl(parsed.query, keep_blank_values=True))
        greenhouse_id = url_posting_id or query.get("gh_jid") or posting_id
        return f"{board}:job:{greenhouse_id}" if board and greenhouse_id else None

    if host == "jobs.ashbyhq.com" or host.endswith(".ashbyhq.com"):
        posting_segment = (
            segments[1]
            if host == "jobs.ashbyhq.com" and len(segments) > 1
            else segments[0]
            if host != "jobs.ashbyhq.com" and segments
            else None
        )
        ashby_id = posting_segment or posting_id
        return f"{board}:job:{ashby_id.lower()}" if board and ashby_id else None

    if host == "jobs.lever.co":
        lever_id = segments[1] if len(segments) > 1 else posting_id
        return f"{board}:job:{lever_id.lower()}" if board and lever_id else None

    if posting_id and host:
        return f"posting:{host}:{posting_id.lower()}"

    normalized_path = parsed.path.rstrip("/").lower()
    if not normalized_path or normalized_path in {"/career", "/careers", "/job", "/jobs"}:
        return None
    return f"url:{canonical_role_url(raw_url)}"


def _hostname(raw_url: str) -> str | None:
    try:
        return (urlsplit(raw_url).hostname or "").lower().removeprefix("www.") or None
    except ValueError:
        return None
